File: python/Day10.py
def part_1() -> None:
    opens: list[str] = ["(", "[", "{", "<"]
    closes: list[str] = [")", "]", "}", ">"]
    incorects: dict[str, int] = {char: 0 for char in closes}

    with open("../data/day10.txt", "r") as dFile:
        for line in dFile.readlines():
            pairs: list[str] = []
            for char in line.strip():
                if char in opens:
                    pairs.append(char)
                elif opens.index(pairs[-1]) == closes.index(char):
                    pairs.pop()
                else:
                    incorects[char] += 1
                    break

    result: int = (
        incorects[")"] * 3
        + incorects["]"] * 57
        + incorects["}"] * 1197
        + incorects[">"] * 25137
    )

    print("Day: 10 | Part: 1 | Result:", result)


def part_2() -> None:
    opens: list[str] = ["(", "[", "{", "<"]
    closes: list[str] = [")", "]", "}", ">"]
    scores: list[list[str]] = []

    with open("../data/day10.txt", "r") as dFile:
        for line in dFile.readlines():
            pairs: list[str] = []
            score: int = 0
            corrupted: bool = False

            for char in line.strip():
                if char in opens:
                    pairs.append(char)
                elif opens.index(pairs[-1]) == closes.index(char):
                    pairs.pop()
                else:
                    corrupted = True
                    break

            if not corrupted:
                while pairs:
                    score = score * 5 + opens.index(pairs.pop()) + 1
                    scores.append(score)

    scores.sort()
    result: int = scores[int(len(scores) / 2)]

    print("Day: 10 | Part: 2 | Result:", result)


if __name__ == "__main__":
    part_1()
    part_2()

File: nodejs/Day10.js
require('./utils.js');

module.exports = {
    this:part1 = function () {
        var opens = ["(", "[", "{", "<"];
        var closes = [")", "]", "}", ">"];
        var incorects = new Map([[")", 0], ["]", 0], ["}", 0], [">", 0]]);

        readFile("../data/day10.txt").forEach((line) => {
            var pairs = [];
            
            for (var index = 0; index < line.length; index++) {
                var char = line.charAt(index);
                if (opens.includes(char))
                    pairs.push(char);
                else if (opens.indexOf(pairs[pairs.length - 1]) == closes.indexOf(char))
                    pairs.pop();
                else {
                    incorects.set(char, incorects.get(char) + 1);
                    break;
                }
            }
        });

        var result = incorects.get(")") * 3 + incorects.get("]") * 57 + incorects.get("}") * 1197 + incorects.get(">") * 25137;
        console.info("Day: 10 | Part: 1 | Result: " + result);
    },

    this:part2 = function () {
        var opens = ["(", "[", "{", "<"];
        var closes = [")", "]", "}", ">"];
        var scores = [];

        readFile("../data/day10.txt").forEach((line) => {
            var pairs = [];
            var score = 0;
            var corrupted = false;
            
            for (var index = 0; index < line.length; index++) {
                var char = line.charAt(index);
                if (opens.includes(char))
                    pairs.push(char);
                else if (opens.indexOf(pairs[pairs.length - 1]) == closes.indexOf(char))
                    pairs.pop();
                else {
                    corrupted = true;
                    break;
                }
            }

            if (corrupted == false)
            {
                while(pairs.length > 0)
                    score = score * 5 + opens.indexOf(pairs.pop()) + 1;
                scores.push(score);
            }
        });

        scores.sort((a, b) => { return a - b; });
        var result = scores[Math.floor(scores.length / 2)];
        console.info("Day: 10 | Part: 2 | Result: " + result);
    }
}
